### ISSUE TITLE
Internal error when selecting invisible LineString

### ISSUE THREAD

**Issue by ewpatton** [mit-cml/appinventor-sources#2441]
**Describe the bug**

[From the forum](https://community.appinventor.mit.edu/t/map-linestring-visibility-bug/28921?u=ewpatton): Selecting a LineString that previously had its visibility set to false will result in an internal error message.

**Affects**

- [x] Designer
- [ ] Blocks editor
- [ ] Companion
- [ ] Compiled apps
- [ ] Buildserver
- [ ] Debugging
- [ ] Other... (please describe)

**Expected behavior**

Selecting an invisible map component should not cause an error. The other components should be investigated to see if they suffer a similar issue.

**Steps to reproduce**

1. Create a map in the designer
2. Create a linestring on the map
3. Uncheck the Visible property
4. Select another component
5. Select the LineString from the component list, an error popup should appear.

**Comment by Himanshu2107** [comment]
I'll try to look into it.

**Comment by PeterMathijssen** [comment]
@Himanshu2107 and @ewpatton this problem also occurs with the polygon, rectangle and circle. 

https://community.appinventor.mit.edu/t/map-linestring-visibility-bug/28921/6?u=peter

**Comment by Himanshu2107** [comment]
The error we get is 
`com.google.gwt.event.shared.UmbrellaException: Exception caught: (TypeError) : Cannot read property 'lastChild' of null`. 
This is caused because some code in the library `leaflet.js` tries to dereference a null object. 
Specifically the native method `setEditing(boolean editing)` (in MockMapFeatureBase.java) calls `feature.bringToFront()`.  `feature` is a JavaScriptObject in MockMapFeatureBase.java. This leads the library code to access `feature._path.parentNode.lastChild`, where `feature._path.parentNode` is null whenever the `Visible` property is set to false. 

Whenever Visible property is changed this method is called:
```
protected native void setNativeVisible(JavaScriptObject map, boolean visible)/*-{
    var feature = this.@com.google.appinventor.client.editor.simple.components.MockMapFeatureBase::feature;
    if (feature) {
      if (visible) {
        map.addLayer(feature);
        feature.getElement().style['pointer-events'] = 'auto';
        feature.getElement().style['cursor'] = 'pointer';
      } else {
        map.removeLayer(feature);
      }
    }
  }-*/;
```
Calling `removeLayer()` causes `feature._path.parentNode` to become `null`. Calling `addLayer()` changes the value to something not null.

So to correct this problem we have to make sure that `feature._path.parentNode` is not null when we change visibility. 
@ewpatton Any ideas?

One thing that I have in mind is that we may not add or remove the layer for changing its visibility, but merely just change its opacity to 1 and 0 respectively. One issue with this approach is that the data from the inactive layer will be loaded, which can be avoided (from [stackoverflow](https://stackoverflow.com/questions/25621307/activate-desactivate-layers-on-leaflet-js)).

**Comment by ewpatton** [comment]
The parent node will naturally be null because when the component is invisible it isn't attached to the map. I think what needs to be done instead is that we shouldn't activate the edit feature when the view is invisible.

**Comment by Himanshu2107** [comment]
@ewpatton
That approach does not work. If we do not activate the edit feature when view is invisible, we would have to enable it whenever the user checks the Visible checkbox. This leads to exactly the same error as before.

**Comment by Himanshu2107** [comment]
I had made the following changes to MockMapFeatureBase.java for this bug:

1. Added instance variable isVisible `protected boolean isVisible = true`.
2. Edited `onSelectedChange(boolean selected)` and `setVisibleProperty(String text)`
```
protected void onSelectedChange(boolean selected) {
    super.onSelectedChange(selected);
    setSelected(selected);
    //for invisible component editing is always disabled
    if (isVisible) {
      setEditing(selected);
    }
  }
```

```
protected void setVisibleProperty(String text) {
    if (map == null) {
      // cannot change marker visibility if there is no map
      return;
    }
    isVisible = Boolean.parseBoolean(text);
    setNativeVisible(map.getMapInstance(), isVisible);
    setEditing(isVisible);
  }
```
These changes seem to solve the bug. But we need that the value of `isVisible` is preserved on project reloads. This is essential because `isVisible` is needed to solve #2444 and the cause of #2444 is hiding this same issue on project reloads. So solving #2444 in some other way will break this code.
(The cause of #2444 is that `addLayer()` is always called on project reloads.) 

@ewpatton How can I make `isVisible` retain its value across project reloads? 

(I think the other instance variables retained there value from the GeoJSON object, but I cannot figure out how to do this for this variable)
